src/main/java/com/puppycrawl/tools/checkstyle/api/AbstractCheck.java

```diff
@@ -278,4 +278,15 @@ public final void log(int lineNo, int colNo, String key,
                 getClass(),
                 getCustomMessages().get(key)));
     }
+
+    /**
+     * The actual context holder.
+     */
+    private static class Context {
```

Comment:
https://github.com/checkstyle/checkstyle/pull/4909#discussion_r131708441
> Also since it contains contents of a file, file violations and file contents, shouldn't it be named something like FileContextHolder?
Also what about AbstractFileSets, does it require a ThreadLocal like this? It contains file information too.

These questions weren't answered.